core/src/main/scala/kafka/log/remote/RemoteLogManager.scala

```diff
@@ -612,23 +616,17 @@ class RemoteLogManager(fetchLog: TopicPartition => Option[Log],
     val offset = fetchInfo.fetchOffset
     val maxBytes = Math.min(fetchMaxBytes, fetchInfo.maxBytes)
 
-    // get the epoch for the requested  offset from local leader epoch cache
-    // FIXME(@kamal), use the epochForOffset API instead of latest epoch.
-    //  val epoch = fetchLog(tp).map(log => log.leaderEpochCache.map(cache => cache.epochForOffset()))
-    var rlsMetadata: Optional[RemoteLogSegmentMetadata] = Optional.empty()
-    fetchLog(tp).foreach { log =>
-      log.leaderEpochCache.foreach(cache => {
-        var epoch = cache.latestEpoch
-        while (!rlsMetadata.isPresent && epoch.isDefined) {
-          rlsMetadata = fetchRemoteLogSegmentMetadata(tp, epoch.get, offset)
-          epoch = cache.findPreviousEpoch(epoch.get)
-        }
-      })
+    val epoch = fetchLog(tp).flatMap(log => log.leaderEpochCache.flatMap(cache => cache.epochForOffset(offset)))
+    val rlsMetadata = if (epoch.isDefined) {
```

Comment:
Nit: In the spirit of keeping the code functional and Scala-like, should we do
```
val rlsMetadata = epoch.flatMap(epoch => fetchRemoteLogSegmentMetadata(tp, epoch, offset))
```

Reply:
+1